### PR TITLE
SNOW-838143 Support PKCS#8 RSA Private Key

### PR DESCRIPTION
**Pull request by sfc-gh-bli**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   *Replace this with a short description of how your code change solves the related issue.*

## Pre-review checklist

(For Snowflake employees)

- [ ] This change has passed precommit
- [ ] I have reviewed code coverage report for my PR in  ([Sonarqube](https://sonarqube.int.snowflakecomputing.com/dashboard?id=snowpark))

### CODE REVIEW

File: src/main/scala/com/snowflake/snowpark/internal/ParameterUtils.scala [hunk 1]
```diff
@@ -106,36 +106,45 @@ private[snowpark] object ParameterUtils extends Logging {
...
-        publicExp,
-        privateExp,
-        prime1,
-        prime2,
-        exp1,
-        exp2,
-        crtCoef)
-      val keyFactory = KeyFactory.getInstance("RSA")
-      keyFactory.generatePrivate(keySpec)
+      val kf = KeyFactory.getInstance("RSA")
```
Comment (sfc-gh-mrui) [inline src/main/scala/com/snowflake/snowpark/internal/ParameterUtils.scala:109]: NIT: Can you add a comment for the function to explain that:
It first attempts to decode as RSA, if failed, fall back to decode as PKSC1

File: src/test/scala/com/snowflake/snowpark/ParameterSuite.scala [hunk 2]
```diff
@@ -65,4 +69,19 @@ class ParameterSuite extends SNTestBase {
       ParameterUtils
         .jdbcConfig(optionWithoutKey + ("privatekey" -> "wrong key"), isScalaAPI = true))
   }
+
+  test("enable to read PKCS#8 private keys") {
```
Comment (sfc-gh-mrui) [inline src/test/scala/com/snowflake/snowpark/ParameterSuite.scala:73]: Can you confirm we have test case to cover for both RSA and PKCS1

File: src/main/scala/com/snowflake/snowpark/internal/ParameterUtils.scala [hunk 3]
```diff
@@ -106,36 +106,45 @@ private[snowpark] object ParameterUtils extends Logging {
...
+      val keySpec = new PKCS8EncodedKeySpec(decoded)
+      kf.generatePrivate(keySpec)
     } catch {
       case e: Exception =>
-        throw ErrorMessage.MISC_INVALID_RSA_PRIVATE_KEY(e.getMessage)
```
Comment (sfc-gh-hachouraria) [inline src/main/scala/com/snowflake/snowpark/internal/ParameterUtils.scala:138]: minor request: Could retain the PKCS#8 attempt's exception's detail in the catch/re-throw when a PKCS#1 attempt also subsequently fails? The original exception's information could be helpful when working with a valid PKCS#8 key that's facing errors when read in Java.

In the current approach we'd just receive a blanket `Could not parse a PKCS1 private key` when passing a PKCS#8 value that's not readable through Java (there are a few algorithms that JSSE [does not support](https://community.snowflake.com/s/article/Private-key-provided-is-invalid-or-not-supported-rsa-key-p8--data-isn-t-an-object-ID)).

File: src/test/scala/com/snowflake/snowpark/ParameterSuite.scala (hunk 2 above)
Comment (sfc-gh-bli) [reply src/test/scala/com/snowflake/snowpark/ParameterSuite.scala:73]: Yes, both `pkcs1` and `pkcs8`

File: src/main/scala/com/snowflake/snowpark/internal/ParameterUtils.scala (hunk 3 above)
Comment (sfc-gh-bli) [reply src/main/scala/com/snowflake/snowpark/internal/ParameterUtils.scala:138]: Sure, great suggestion. I changed it to print error messages of both `pkcs1` and `pkcs8`.

File: src/main/scala/com/snowflake/snowpark/internal/ParameterUtils.scala (hunk 1 above)
Comment (sfc-gh-bli) [reply src/main/scala/com/snowflake/snowpark/internal/ParameterUtils.scala:111]: added

**Review by sfc-gh-mrui — APPROVED** [review]
LGTM

**Pull request merged**